Fix #1442 and #3292: Answer migration and summarizers framework

BenHenning opened the pull request:
The epic culmination and finale of #1171 and #1205. This PR aims to perform some minor iterations on top of the work done in #1171 and #1205 to make this change production ready. Namely, the following needs to be completed (in order) before this can be merged into develop:

- [x] Bring this branch up-to-date with develop and resolve any conflicts
- [x] Resolve all linting issues
- [x] Verify all tests are working and passing (and fix any that are currently broken)
- [x] Write down the playbook for performing the migration using notes from #1205 and include them in one location (in this PR)
- [x] Play through a bunch of explorations when answers are saved to the new model and make sure everything seems to be working properly (esp. in regards to stats and the continuous stats jobs)
- [x] Verify old jobs do not break with the changes to job output logic
- [x] Re-run the job in a production environment on the full answer data set as a final sanity check before merging this into develop
- [x] Post production migration, perform a similar sanity check as mentioned above (answer visualizations and various checks) to ensure everything is looking good
- [x] Resolve #3293

Once all the above are completed, this PR should have a final sanity check before being merged into develop.

BenHenning commented:
In retrospect the first checkbox doesn't make much sense:

> Implement timestamps for submitted answers (these cannot be reconstituted, so store the timestamp at which the migration happened)

We don't actually store timestamps for answers, just time duration. Removing this item.

BenHenning commented:
## Answer Migration Instructions

1. Go into the admin panel
2. Stop all continuous computation jobs and wait for them to stop
2. Run ``ClearMigratedAnswersJob`` and ``ClearLargeAnswerBucketsJob``
3. Run the ``PurgeInconsistentAnswersJob`` and wait for it to complete
4. Run the ``SplitLargeAnswerBucketsJob``
5. Run the ``AnswersAudit`` job
6. Wait for both jobs to complete
7. Run the ``AnswerMigrationJob`` and wait for it to complete
8. Check the output of the ``AnswerMigrationJob`` for any errors.  If there are any errors follow these sub-steps:
  a. Run ``AnswerMigrationCleanupJob`` and wait for it to complete
  b. Re-run the ``AnswerMigrationJob`` and wait for it to complete
  c. Check the output of the ``AnswerMigrationJob`` for any errors. If the errors are the same as the previous time the ``AnswerMigrationJob`` was run, stop the process and investigate.  Otherwise, repeat steps (a) through (c) until the migration job no longer reports errors.
9. Start the ``AnswersAudit2`` job
10. Start the ``RuleTypeBreakdownAudit`` job
11. Wait for both to complete.  Check the output of ``RuleTypeBreakdownAudit`` to verify that all answer rule types were migrated (there should be no reports of specific exploration failing to migrate all of their rule types)
12. Optionally compare the output from ``AnswersAudit`` and ``AnswersAudit2``.  Other than some minor differences between the old and new storage models, the aggregate rule type counts should be equal for both jobs
13. Restart continuous computation jobs and start ``InteractionAnswerSummariesAggregator``
14. After the aggregator has finished running a few times, look through the answer stats for some large explorations and verify they look similar to before

BenHenning commented:
Added job for clearing the interaction registry cache for after the answer migration job since the answer migration job permanently injects two faux interactions into the registry cache.

BenHenning commented:
When verifying behavior in production a final time, a number of problems occurred.

The first problem was when we ran the answer migration job. At one point during the development process, we lowered the accumulated submitted answer list size to 100,000 bytes, whereas it intended to be 1,000,000 bytes. After restoring it to 1m bytes in the review, however, running the job in production led to ``put_multi`` calls exceeding the datastore size limits for an entity. Further investigation yielded that some entities were indeed exceeding 1MB in size, so the model was updated to only allow accumulated answer lists to 900,000 bytes to allow plenty of padding for metadata and other, unaccounted for data that needs to be stored in entities (124,000 bytes of padding).

Second, some answers were not properly migrated:

``Item ID: 13.First State.submit.IsInRegion(9), error: Expected IsInRegion occurrence to refer to the region actually containing the clicked coordinates (2 times)``

This only happened to 6 buckets of answers (14 specific answers) so it's probably fine just to ignore these answers and migrate them anyway. We can do work later on to re-validate all present answers in ``StateAnswersModel``, but there isn't a good solution here. Instead, all the regions the coordinates of those 14 answers overlap are now being stored after the migration, in addition to the one indicated by the rule spec (the one that the answer was actually classified to contain). This seems like a reasonable compromise given the data.

The third issue is with the interaction summarizers continuous computation job. When we tried running the job on the backup migration production server after running the answer migration, we saw a number of MR pipeline failures:

``Error: Buffer too big. Can't write more than 1024000 bytes in one request: risk of writes interleaving. Got: 1048576``

These happened toward the end of the shuffle-hash step of the MR pipeline. Some investigation into app engine yielded this was due to the value part of the yielded map pair exceeding 1 MB, which seems really likely since the interation summarizers job outputs an aggregated list of submitted answers for all versions of a given exploration and state. This has been updated to output the ``StateAnswersModel`` item IDs instead of the answers themselves, then the models are looked up efficiently in the reduce step of the job. The memory consumption requirement is the same in reduce() and the number of datastore ops probably will go up, but the pressure on the MR pipeline has drastically gone done, which led to the MR job being much more efficient and not having the failure.

Finally, we noticed some inconsistencies in the aggregated answer stats before and after the migration. We will need to come to a decision about these inconsistencies before merging in the job.

BenHenning commented:
Thanks to @seanlip for resolving #3293!

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -1046,6 +1046,26 @@ def reduce(key, stringified_values):
+    @staticmethod
+    def reduce(key, stringified_values):
+        # pylint: disable=protected-access
+        interaction_registry.Registry._refresh()
```
It seems incredibly heavyweight to do this using an MR job. Why not just click a button in the admin dashboard that performs this action? I don't understand the point of iterating through all the explorations...

BenHenning replied:
Ah, yes an admin button makes more sense (for some reason I didn't think of that at the time...) Nevertheless, this will only be run 2 more times (once for a test, and once for the real migration). Adding the additional button in the admin panel involves quite a bit more than just keeping this job here. It is heavyweight, but it's not too bad and it runs quickly.

seanlip commented on core/domain/stats_jobs_one_off.py:
```diff
@@ -1046,6 +1046,26 @@ def reduce(key, stringified_values):
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    # pylint: disable=unused-argument
```
Put this on the right of line 1057.

BenHenning replied:
I'm also using it to disable unused arguments in reduce(), but I added it to both lines individually instead of using a block disable.

seanlip commented on core/storage/statistics/gae_models.py:
```diff
@@ -603,9 +603,11 @@ class StateAnswersModel(base_models.BaseModel):
-    _MAX_ANSWER_LIST_BYTE_SIZE = 1000000
+    # This provides about 124k of padding for the other properties and entity
+    # storage overhead (since the max entity size is 1MB). The meta data can
+    # get close to 50k or exceed it, so plenty of padding is leftover to avoid
```
Nit: leftover --> left

BenHenning replied:
Done.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -545,56 +543,49 @@ def _encode(item):
-        interaction_id = ast.literal_eval(
-            stringified_state_answers_dicts[0])['interaction_id']
+    def reduce(key, stringified_item_ids):
+        exploration_id, exploration_version, interaction_id, state_name = (
```
Just to check, can the state name contain colons?

Also, is exploration_version _supposed_ to be a string?

BenHenning replied:
The version can be a string here (see VERSION_ALL at the top of the file).

Per ``feconf``:

```python
# A string containing the disallowed characters in state or exploration names.
# The underscore is needed because spaces in names must be converted to
# underscores when displayed as part of a URL or key. The other conventions
# here are derived from the Wikipedia guidelines for naming articles.
INVALID_NAME_CHARS = u':#/|_%<>[]{}\ufffd\\' + chr(127)
for ind in range(32):
    INVALID_NAME_CHARS += chr(ind)
```

Colons are not allowed in state names, so this split should be fine. I usually prefer extracting the values from the value dicts instead, but that significantly increases the amount of memory passed through the pipeline in this case.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -545,56 +543,49 @@ def _encode(item):
-        # memory. If you came across this TODO when investigating a shard
-        # failure for this job due to exceeding the soft private memory limit,
-        # this is probably the reson.
+        # Retrieve all associates IDs to StateAnswersModel.
```
Retrieve the IDs of all associated StateAnswersModel entities.

BenHenning replied:
Done (reworded it during some of the alterations to this job).

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -545,56 +543,49 @@ def _encode(item):
+
+        # Retrieve all StateAnswerModel entities associated with this shard
+        # instance.
+        state_answer_models = stats_models.StateAnswersModel.get_multi(item_ids)
```
state_answers_models?

BenHenning replied:
Done.

seanlip commented:
@BenHenning I've done the review. Thanks!

BenHenning commented:
Updated the answer summarizers to, when aggregating stats across all versions of an exploration/state, to not include answers submitted to versions whose interaction ID have changed from the latest version (since those answers are no longer meaningful). Improved some error handling when answer calculation output fails to save to the data store and addressed the remaining review comments.

So far when running the newly changed summarizer job in prod, the stats still seem broken in precisely the same way. I'm currently investigating why.

BenHenning commented:
The working theory is that the aggregated stats look different before and after the migration because there isn't a linear mapping from the old to new answer model. The old model maps answers to rule spec strings, which depend both on interactions and changes to the rules of states. They do not depend on versions. The new storage model stores answers only based on exploration IDs, state names, and versions (which is more granular than both interaction IDs and rule spec strings, since individual exploration versions are immutable--each change to a state's interaction or rules will create new exploration versions).

However, the problem comes from aggregating answers across all versions of an exploration. The new storage model provides a much larger bucket of answers of answers to aggregate from. There are two identifiable cases where there will be more answers available to the summarizers job than previously accessed by the exploration editor before the migration:

1. Across interaction ID changes (where the interaction also has different rule types)
2. Across rule changes

The former is likely less common, but it was addressed earlier when the summarizers job was updated to only count answers up to versions with changed interactions, but not beyond those versions. The latter still poses a discrepancy, however. If the answer 122 matched to one rule spec before, and another rule spec now, all answers of value 122 submitted to both rules will be counted together as 122 by the new interaction summarizers job, whereas the old exploration editor would only show the most recent 122 values since the older rule no longer exists (so the corresponding answers cannot be retrieved unless a similar rule were reintroduced).

This generally means the new answer summarizer job will generally yield more answers, as more are available. This isn't wrong (as the latest job performing answer value validation for exploration 0's NumericInput state seems to suggest), it's simply a different way of looking at the available data (and makes more data available). Explorations have had more answers submitted to them than previously suggested and this data will be brought out with this new framework.

Given that analysis and the small confirmation we needed with the new validation job, I think we're at a point where we're comfortable with the answer migration now running for oppia.org. This PR will be wrapped up and merged into develop momentarily.

BenHenning commented:
Post migration validation has been completed as a trial run on the backup migration server. One more will be done immediately prior to the migration as part of the actual migration process.

The item tracking updating the release wiki page is being tracked at a higher level than this PR, so it will be removed from here and not gate the submission of this PR.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -526,41 +526,118 @@ def entity_classes_to_map_over(cls):
+        latest_version = versions[0]
+        latest_interaction_id = versioned_interaction_ids[latest_version][0]
+        # NOTE TO DEVELOPER: This skips the first version in the list since that
+        # interaction ID is already being compraed. Bear in mind that skipping
```
nit: compraed --> compared

BenHenning replied:
Done.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -526,41 +526,118 @@ def entity_classes_to_map_over(cls):
+        # the map function). For this reason, all of this code is simply skipped
+        # in that case in favor of performance.
+        if len(versions) > 1:
+            start_version_index = next(
```
Convoluted. Strongly recommend breaking down into multiple statements.

"start_version" doesn't really convey what you want IMO. Maybe earliest_version_with_current_interaction_id

Or actually -- earliest_acceptable_version may be nicer. (Shorter, and still conveys more than "start".)

BenHenning replied:
Done.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -526,41 +526,118 @@ def entity_classes_to_map_over(cls):
+            for ignored_version in ignored_versions:
+                del versioned_interaction_ids[ignored_version]
+                del versioned_item_ids[ignored_version]
+            versions = versions[0:start_version_index+1]
```
Note that the 0 can be omitted.

BenHenning replied:
Done.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -526,41 +526,118 @@ def entity_classes_to_map_over(cls):
+            versioned_key = u'%s:%s:%s' % (
+                item.exploration_id, item.exploration_version, item.state_name)
+            yield (versioned_key.encode('utf-8'), {
+                'item_id': item.id,
```
"item" is really too vague, and I keep forgetting what item we are talking about. Can we call it state_answers_model_id perhaps?

BenHenning replied:
Done.

seanlip commented on core/domain/stats_jobs_continuous_test.py:
```diff
@@ -637,6 +637,151 @@ def test_answers_across_multiple_exploration_versions(self):
+                'new_value': 'TextInput',
+            }], 'Change to TextInput')
+
+            # Submit another numeber-like answer.
```
typo: number

BenHenning replied:
Fixed.

BenHenning commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -526,41 +526,118 @@ def entity_classes_to_map_over(cls):
+            for value_dict in value_dicts]))
+        versions.sort(reverse=True)
+
+        # In half of the non-aggregated outputs from the map function, versions
```
Super close to what I wrote, but I took this instead. It was slightly clearer.

BenHenning commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -526,41 +526,118 @@ def entity_classes_to_map_over(cls):
+            yield (
+                'Expected a single version when aggregating answers for '
+                'exploration %s (v=%s), but found: %s' % (
+                    exploration_id, exploration_version, versions))
```
Seems reasonable, done.

BenHenning commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -526,41 +526,118 @@ def entity_classes_to_map_over(cls):
+        versioned_item_ids = {version: set() for version in versions}
+        for value_dict in value_dicts:
+            version = value_dict['exploration_version']
+            versioned_interaction_ids[version].update(
```
Fatigue and lack of memory. :)

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -617,20 +618,23 @@ def reduce(key, stringified_values):
-                     != latest_interaction_id)),
-                len(versions) - 1)
-            start_version = versions[start_version_index]
+            indexes = [
```
invalid_version_indexes

BenHenning replied:
Done.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -617,20 +618,23 @@ def reduce(key, stringified_values):
+                if versioned_interaction_ids[version][0] != (
+                    latest_interaction_id)]
+            earliest_acceptable_version_index = (
+                indexes[0] if indexes else len(versions) - 1)
```
OK -- the -1 is indeed terrible.

How about:

    invalid_version_indexes = [index for index, version in enumerate(versions) if ...]
    earliest_acceptable_version_index = (invalid_version_indexes[0] - 1 if indexes else len(versions) - 1)

BenHenning replied:
Haha, that seems reasonable. Done.

BenHenning commented:
Please refrain from merging this in. I will merge in it once ready.

BenHenning commented:
And thanks @seanlip for the massive amount of reviewing and back-and-forth discussions you provided over the last 16 months of this project.

seanlip commented:
:D

BenHenning commented:
We are going to merge this in as a standard merge due to the large number of commits. We want to preserve history and avoid complicated edge cases that might come up during a squash + rebase.
